Issue/12563 delete custom fields

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12563 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds support for deleting custom fields.

### Steps to reproduce
1. Open wp-admin, and add at least one custom field to an order.
2. Open the order in the app.
3. Tap on "View custom fields"
4. Open an existing custom field.
5. Tap on the overflow menu button.
6. Tap on Delete button.

### Testing information
1. Test the delete flow for an existing field.
2. Test the delete flow for a new non-saved field.
3. Test the undo SnackBar.
4. Test saving changes after deleting some items.

### The tests that have been performed
Same as above.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Bump fluxc

hichamboushaba committed:
Add initial logic of deletion

hichamboushaba committed:
Allow setting a custom color for overflow menu items

hichamboushaba committed:
Handle deleting fields

hichamboushaba committed:
Hide the delete button when creating a new item

hichamboushaba committed:
Show an undo snackbar after deleting a field

hichamboushaba committed:
Add unit tests

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/235">20.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Use Compose for displaying Snackbars instead of UIMessageResolver

This allows having the Snackbar on the correct location above the FAB, instead of hiding it.

hichamboushaba committed:
Bump fluxc

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorScreen.kt:
```diff
@@ -60,6 +64,24 @@ private fun CustomFieldsEditorScreen(
                     }
+                    if (!state.isCreatingNewItem) {
+                        WCOverflowMenu(
+                            items = listOf(R.string.delete),
```
We'll add more items in a future PR to support copying the key and value.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/list/CustomFieldsFragment.kt:
```diff
@@ -4,35 +4,37 @@ import android.os.Bundle
 
-    @Inject
-    lateinit var uiMessageResolver: UIMessageResolver
+    private val snackbarHostState = SnackbarHostState()
```
I updated the logic to show the Snackbars using Compose instead of the previous approach, this is needed to make sure the `Snackbar` doesn't overlap with the FAB button, I'm not aware of any way to make this work well with the previous approach.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorFragment.kt:
```diff
@@ -36,7 +32,7 @@ class CustomFieldsEditorFragment : BaseFragment() {
         viewModel.event.observe(viewLifecycleOwner) { event ->
             when (event) {
-                is MultiLiveEvent.Event.ExitWithResult<*> -> navigateBackWithResult(RESULT_KEY, event.data)
+                is MultiLiveEvent.Event.ExitWithResult<*> -> navigateBackWithResult(event.key!!, event.data)
```
I'm sure you considered not using `!!`, but I don't see any harm in using:
```
 event.key?.let { key -> navigateBackWithResult(key, event.data) } 
```
This could help prevent potential crashes in the future.

hichamboushaba replied:
Thanks @irfano, but personally, I prefer the app to crash than to not behave as expected (offensive programming), and for this case, unless someone changes the code, the `key` will always be non-null, and if someone changes the code, it would be good to have a crash that will help fix the issue quickly than to have it go unnoticed.
WDYT?

irfano replied:
Makes sense! Offensive programming seems appropriate for this case.

irfano reviewed (commented):
LGTM! 👍🏻 Works well! I added a comment and will approve and merge after you respond to it.

irfano reviewed (approved):
Good job! LGTM! You'll probably do, but as a reminder, don't forget to add RELEASE NOTES in one of your PRs for the custom fields feature.

hichamboushaba commented:
> You'll probably do, but as a reminder, don't forget to add RELEASE NOTES in one of your PRs for the custom fields feature.

Thanks @irfano for the reminder, what we do generally when working on a feature that's behind a feature flag, is that we add the release notes entry in the PR that enables or removes the feature flag, so that it's not announced before we release it.

hichamboushaba committed:
Merge branch 'trunk' into issue/12563-delete-custom-fields
